### PR TITLE
rewrite: Use r.URL.RawPath (#3596)

### LINKED ISSUE #3596: v2: Use r.URL.RawPath for rewrite
When using e.g. `uri replace` paths get automatically unescaped. This might be undesirable. Something like this would fix it:

```diff
diff --git a/modules/caddyhttp/rewrite/rewrite.go b/modules/caddyhttp/rewrite/rewrite.go
index d47c388a..4370a77f 100644
--- a/modules/caddyhttp/rewrite/rewrite.go
+++ b/modules/caddyhttp/rewrite/rewrite.go
@@ -303,13 +303,24 @@ func (rep replacer) do(r *http.Request, repl *caddy.Replacer) bool {
        find := repl.ReplaceAll(rep.Find, "")
        replace := repl.ReplaceAll(rep.Replace, "")
 
-       oldPath := r.URL.Path
+       oldPath := r.URL.RawPath
        oldQuery := r.URL.RawQuery
 
-       r.URL.Path = strings.Replace(oldPath, find, replace, lim)
        r.URL.RawQuery = strings.Replace(oldQuery, find, replace, lim)
 
-       return r.URL.Path != oldPath && r.URL.RawQuery != oldQuery
+       if oldPath == "" {
+               oldPath = r.URL.Path
+               r.URL.Path = strings.Replace(oldPath, find, replace, lim)
+
+               return r.URL.Path != oldPath && r.URL.RawQuery != oldQuery
+       }
+
+       r.URL.RawPath = strings.Replace(oldPath, find, replace, lim)
+       if p, err := url.PathUnescape(r.URL.RawPath); err == nil {
+               r.URL.Path = p
+       }
+
+       return r.URL.RawPath != oldPath && r.URL.RawQuery != oldQuery
 }
 
 // Interface guard
```

What do you think? Maybe there's a better way?

### PR DESCRIPTION
**Pull request by go-d**
Prevent information loss, i.e. the encoded form that was sent by the
client, when using URL strip/replace.

Fixes #3596

### CODE REVIEW

**Comment by mholt** [comment]
Awesome, thanks!! Will look at getting this in shortly after the 2.3 release.

**Comment by go-d** [comment]
There is an edge case that we still need to consider:
```go
rule:   Rewrite{URISubstring: []replacer{{Find: "findme", Replace: "replaced"}}},
input:  newRequest(t, "GET", "/foo%2findme/bar"),
expect: newRequest(t, "GET", "/foo%2findme/bar"),
```
Using a regular expression in place of `strings.Replace` could be the start of a solution:
```go
re := regexp.MustCompile("(^|^[^%]|[^%][^%])" + regexp.QuoteMeta(find))
r.URL.RawPath = re.ReplaceAllString(r.URL.RawPath, "${1}"+replace)
```
Or maybe a custom version of strings.Replace?

**Comment by mholt** [comment]
Definitely not a regular expression... what is the problem there exactly?

**Comment by go-d** [comment]
The problem is the possibility of unexpected results. A search string having a-f or 0-9 as the first character might affect percent-encoded (%hh) bytes, like so:

1. replace *findme* with *replace*, giving *%2r*... (invalid hex)
2. replace *findme* with *0*, giving *%20* (unintended space)

A replacer that is aware of %-encoding would not alter the URL in both these cases, instead of the currently unpredictable */foo/indme/bar* and */foo%20/bar*, respectively.

**Comment by mholt** [comment]
Oh; I see. So we just need to do the replacement on the decoded form of the URL, rather than the encoded form.

**Comment by go-d** [comment]
I don't think so. How would you then reconstruct the encoded form (RawPath)?

My regexp example works, but could certainly be done without regexp.

**Comment by mholt** [comment]
(Sorry for the delay, will get around to this shortly after the 2.3 release in a few days!)

**Comment by mholt** [comment]
Ok after thinking about this more, and fiddling with it for some time today, I don't think that edge case is a problem.

The rewrite module works on the raw URL - this allows you to do replacements with `%20` versus `/` explicitly. And users can already write their own regexes.

So I think that test case is incorrect. What the rewrite module currently does (with this change) is correct.

Ready for final review?

**Comment by go-d** [comment]
Sure.

Hard to say what is correct though, since the behavior isn't specified.

Agreed that it is likely not much of a problem in practice.

A modified `strings.Replace` that looks behind (in case search string 
starts with a hex digit) to ensure no `%` is there would make the result 
easier to predict in my opinion.

That would require some copying from the standard library. Of code that 
rarely changes. But still.

**Comment by mholt** [comment]
That's essentially trying to parse the _decoded_ form of the path, not the RawPath which we're doing now. If we leave it raw/encoded, we don't lose information. Giving a final review soon. Thanks!

**Comment by go-d** [comment]
Maybe looking for `%` in RawPath and RawQuery (before a string match index) could be considered a form of parsing. But that seems inevitable if one wants to guard against incorrect partial replacements of hex values therein.

At no point is decoding happening there, or URL.Path used for that, though. So no information loss.

For a future issue, perhaps.

**Review by mholt — APPROVED** [review]
> Maybe looking for % in RawPath and RawQuery (before a string match index) could be considered a form of parsing. But that seems inevitable if one wants to guard against incorrect partial replacements of hex values therein.

Right; the thing is, if we don't want the encoded characters to be rewritten we should be operating on `Path` like we are now, which is decoded, but the problem is it's harder to go from Path to RawPath. As your PR shows, it's much easier to go from RawPath to Path.

> At no point is decoding happening there, or URL.Path used for that, though. So no information loss.

Oh, yeah, haha -- exactly.

Let's roll with this for a while, I think this is better as-is, but we'll see if we get any complaints. Still, I think even those can be addressed with something like "do the rewrite differently."

Basically, we'll expect the user to configure correct rewrites -- and of course, clients need to continue making correct requests with correct URLs.

**Comment by go-d** [comment]
Thank you for creating Caddy!

(Commit message now says "instead of Path", but usually, RawPath is empty, and Path is used.)
